client/service/src/metrics.rs

```diff
@@ -0,0 +1,384 @@
+
+use crate::NetworkStatus;
+use netstat2::{TcpState, ProtocolSocketInfo, iterate_sockets_info, AddressFamilyFlags, ProtocolFlags};
+use prometheus_endpoint::{register, Gauge, U64, F64, Registry, PrometheusError, Opts, GaugeVec};
+use sc_client::ClientInfo;
+use sc_telemetry::{telemetry, SUBSTRATE_INFO};
+use std::convert::TryFrom;
+use sp_runtime::traits::{NumberFor, Block, SaturatedConversion, UniqueSaturatedInto};
+use sp_transaction_pool::PoolStatus;
+use sp_utils::metrics::register_globals;
+use sysinfo::{ProcessExt, System, SystemExt};
+
+#[cfg(not(unix))]
+use sysinfo::get_current_pid;
+
+#[cfg(unix)]
+use procfs;
+
+struct PrometheusMetrics {
+	// system
+	load_avg: GaugeVec<F64>,
+
+	// process
+	cpu_usage_percentage: Gauge<F64>,
+	memory_usage_bytes: Gauge<U64>,
+	netstat: GaugeVec<U64>,
+	threads: Gauge<U64>,
+	open_files: GaugeVec<U64>,
+
+	// -- inner counters
+	// generic info
+	block_height: GaugeVec<U64>,
+	number_leaves: Gauge<U64>,
+	ready_transactions_number: Gauge<U64>,
+
+	// I/O
+	network_per_sec_bytes: GaugeVec<U64>,
+	database_cache: Gauge<U64>,
+	state_cache: Gauge<U64>,
+	state_db: GaugeVec<U64>,
+}
+
+impl PrometheusMetrics {
+	fn setup(registry: &Registry, name: &str, version: &str, roles: u64)
+		-> Result<Self, PrometheusError>
+	{
+        register(Gauge::<U64>::with_opts(
+            Opts::new(
+                "build_info",
+                "A metric with a constant '1' value labeled by name, version"
+            )
+                .const_label("name", name)
+                .const_label("version", version)
+		)?, &registry)?.set(1);
+		
+        register(Gauge::<U64>::new(
+            "node_roles", "The roles the node is running as",
+		)?, &registry)?.set(roles);
+
+		register_globals(registry)?;
+		
+		Ok(Self {
+
+			// system
+			load_avg: register(GaugeVec::new(
+				Opts::new("load_avg", "System load average"),
+				&["over"]
+			)?, registry)?,
+
+			// process
+			memory_usage_bytes: register(Gauge::new(
+				"memory_usage_bytes", "Node memory usage",
```

Comment:
Why remove the information that it is the resident set size?

(See master version: https://github.com/paritytech/substrate/blob/master/client/service/src/builder.rs#L80)